### ISSUE TITLE
Tests for Decision_list_learner

### ISSUE THREAD

**Issue by CharuChhimpa** [aimacode/aima-python#856]
@norvig and @delaray, tests for Decision_list_learner are not added in test_learning.py. Can I add that?

**Comment by MrDupin** [comment]
The Decision List Learner is not yet complete, there are some details missing from its implementation. These need to be finished first, and then we can add the tests.

If you want you can take this up.

**Comment by CharuChhimpa** [comment]
Ok. So, anyone working on implementing Decision List learner? Can I start on it?

**Comment by ad71** [comment]
I don't think anyone is working on this. You can go ahead.

**Comment by blue-atom** [comment]
@CharuChhimpa If you decide to quit work on this, please let me know.

**Comment by CharuChhimpa** [comment]
@ad71, Can I know why I am not visible in the contributers page. As my previous commit is already merged.

**Comment by kunwar31** [comment]
@CharuChhimpa Sometimes it can take a few hours for the contributors page to update. I noticed the same problem when my first pull request was merged.

**Comment by ad71** [comment]
@CharuChhimpa I think @kunwar31 is right, but I'm pretty sure the number of contributors has updated. It says 74, but there are an odd number of contributors in the contributors page. Have you deleted your fork of the repository after your commit got merged? If so, your name won't show up on the contributors list.

**Comment by CharuChhimpa** [comment]
@ad17, I have not deleted my fork.

**Comment by ad71** [comment]
Then maybe you should wait for a few hours and see if it updates, because I have never seen this happen before.

**Comment by CharuChhimpa** [comment]
@ad17, Can you please tell me what does the test signifies in the decision list learner?

**Comment by ad71** [comment]
I'm sorry I didn't understand your question. There are currently no tests for decision list learner. We need simple assertion-based tests that show that the algorithm actually works and produces the same output every time. I believe using the iris dataset is advisable as most tests for learners are written for this dataset. Check the source code of a few existing tests to see what is required.

**Comment by CharuChhimpa** [comment]
No, I am not talking about the tests. I am talking about the test variable that is used in the algorithm of Decision List Learner .

**Comment by ad71** [comment]
Oh, `test` here refers to any input given to the decision list learner. At first glance, it seems that the `DecisionListLearner` class was never completed or it is meant to be a base class for other classes. `find_examples` is supposed to find a set of examples that give the same outcome (output) for a given test (input). It is supposed to return the `test`, the `outcome` and the examples that it found, but this function doesn't exist as of now.
The `passes` function is supposed to take an `example` and apply the `test` to all its elements and return true if all of them return the same `outcome` and false otherwise. This function is missing too. I'm not sure if these functions were meant to be completed or this class was meant to be a base class. Maybe @MrDupin and @norvig can help.

**Comment by CharuChhimpa** [comment]
@norvig and @MrDupin, Can you tell me that the functions find_examples and passes in decision list learner needs to be implemented?

**Comment by MrDupin** [comment]
I cannot tell, sorry. I am only a student. :)

**Comment by CharuChhimpa** [comment]
@norvig, Do we need to implement 'find_examples' and 'passes' functions in Decision list leanrner?

**Comment by CharuChhimpa** [comment]
@norvig, Please tell me that the functions 'find_examples' and 'passes' in DecisionListLearner needs to be implemented or not?

**Comment by pyaf** [comment]
Hey guys, I've looked into `DecisionListLearner` as mentioned by @CharuChhimpa `find_examples` and `passes` are not implemented (or it is because it is meant to be a Base class for other classes?).

I would like to complete `DecisionListLearner` as mentioned in the pseudocode. Should I go ahead with this?

**Comment by MrDupin** [comment]
@pyaf: Yes, that would be awesome!

**Comment by alannair** [comment]
Hi!
I'm new here, but I'd like to handle this issue as well. 
To start with, I see in the `learning.ipynb` file, there is a reference to a method `pseudocode` that should return the pseudocode of Decision Tree Learning, but this method seems to be missing in the `learning.py` file. Shall I add that?

**Comment by dmeoli** [comment]
Done together with the `DecisionListLearner` implementation (#1048): added `test_decision_list_learner` to `tests/test_learning.py` and `tests/test_learning4e.py`, which train the learner on the discrete `restaurant` dataset and assert it classifies every training example correctly (0 training error) and that the list ends in the empty catch-all test. Closing as completed.
